network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go

```diff
@@ -0,0 +1,180 @@
+	require.Zerof(t, gauge2, "expected second gauge to be 0")
+	require.Equal(t, uint(2), cache.c.Size(), "expected cache to have two additional records")
+}
+
+// TestDuplicateMessageTrackerCache_ConcurrentInit tests the concurrent initialization of records.
+// The test covers the following scenarios:
+// 1. Multiple goroutines initializing records for different node IDs.
+// 2. Ensuring that all records are correctly initialized.
+func TestDuplicateMessageTrackerCache_ConcurrentInit(t *testing.T) {
+	cache := duplicateMessageTrackerCacheFixture(t, 100, defaultDecay, defaultSkipDecayThreshold, zerolog.Nop(), metrics.NewNoopCollector())
+
+	peerIDs := unittest.PeerIdFixtures(t, 10)
+
+	var wg sync.WaitGroup
+	wg.Add(len(peerIDs))
+
+	for _, peerID := range peerIDs {
+		go func(id peer.ID) {
+			defer wg.Done()
+			gauge, found, err := cache.GetWithInit(id)
+			require.NoError(t, err)
+			require.True(t, found)
+			require.Zerof(t, gauge, "expected all gauge values to be initialized to 0")
+		}(peerID)
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+}
+
+// TestDuplicateMessageTrackerCache_ConcurrentSameRecordInit tests the concurrent initialization of the same record.
+// The test covers the following scenarios:
+// 1. Multiple goroutines attempting to initialize the same record concurrently.
+// 2. Only one goroutine successfully initializes the record, and others receive false on initialization.
+// 3. The record is correctly initialized in the cache and can be retrieved using the GetWithInit method.
+func TestDuplicateMessageTrackerCache_ConcurrentSameRecordInit(t *testing.T) {
+	cache := duplicateMessageTrackerCacheFixture(t, 100, defaultDecay, defaultSkipDecayThreshold, zerolog.Nop(), metrics.NewNoopCollector())
+
+	peerID := unittest.PeerIdFixture(t)
+	const concurrentAttempts = 10
+
+	var wg sync.WaitGroup
+	wg.Add(concurrentAttempts)
+
+	for i := 0; i < concurrentAttempts; i++ {
+		go func() {
+			defer wg.Done()
+			gauge, found, err := cache.GetWithInit(peerID)
+			require.NoError(t, err)
+			require.True(t, found)
+			require.Zero(t, gauge)
+		}()
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// ensure that only one goroutine successfully initialized the record
+	require.Equal(t, uint(1), cache.c.Size())
+}
+
+// TestDuplicateMessageTrackerCache_DuplicateMessageReceived tests the DuplicateMessageReceived method of the RecordCache.
+// The test covers the following scenarios:
+// 1. Updating a record gauge for an existing peer ID.
+// 2. Attempting to update a record gauge  for a non-existing peer ID should not result in error. DuplicateMessageReceived should always attempt to initialize the gauge.
+// 3. Multiple updates on the same record only initialize the record once.
+func TestDuplicateMessageTrackerCache_DuplicateMessageReceived(t *testing.T) {
```

Comment:
A concurrent version of this test will be highly valuable.

Reply:
https://github.com/onflow/flow-go/pull/5058/commits/91a1a63836ca1b5a3e70065ec022fcc07397d03a